src/main/java/org/truffleruby/language/LexicalScope.java

```diff
@@ -54,6 +59,19 @@ public static DynamicObject resolveTargetModuleForClassVariables(LexicalScope le
         return scope.liveModule;
     }
 
+    public Map<DynamicObject, DynamicObject> getRefinements() {
+        return refinements;
+    }
+
+    public Map<DynamicObject, DynamicObject> getRefinementsAllScopes() {
+        Map<DynamicObject, DynamicObject> results = new HashMap<>();
+        results.putAll(refinements);
+        if (this.getParent() != null) {
+            results.putAll(this.getParent().getRefinementsAllScopes());
```

Comment:
This creates `N` maps and does O(n^2) `put()`.
Could you use iteration or a separate getRefinementsAllScopes(Map<> map) to avoid creating extra maps?